### ISSUE TITLE
JCR Migration: Upgrade HttpClient from 3.x to 4.x

### ISSUE THREAD

**Issue by TheAntimist** [caldav4j/caldav4j#63]
I had been waiting for the Jackrabbit Webdav upgrade for 4.x to happen, now that finally [JCR-2406](https://issues.apache.org/jira/browse/JCR-2406) is closed, perhaps for the next release we can look into updating this.

**Comment by ioggstream** [comment]
LOOL
```
Created:	24/Nov/09 13:30
Updated:	12/Jan/17 12:47
Resolved:	30/Dec/16 14:39
```

**Comment by TheAntimist** [comment]
Yup, it's one of the longer issues, 8 years is a long time. Anyway, regarding the next release I'm thinking of releasing the current one i.e. with HttpClient 3.x with docs.

For the next release we'll add the HttpClient 4.x, only thing is do we make sure that we want backward compatibility with 3.x.

**Comment by ioggstream** [comment]
Yes, first things first!

**Comment by ioggstream** [comment]
@alexander233 Do your branch work with http4? https://github.com/caldav4j/caldav4j/compare/master...alexander233:httpclient4-dev

**Comment by alexander233** [comment]
Yes, it does. I am still working on it; I think I have gotten most of the functionality migrated. But I am still working on the testing side. I am already using my branch in one application, but I have not yet gotten around to testing it with Google.

**Comment by ioggstream** [comment]
@TheAntimist when you've time can you please check if it's mergeable?
@alexander233 it would be very nice if you manage to:

  - rebase on the latest master 
  - minimize the diff (eg. just make functional changes and not stylistic ones) 
  
A possible way could be:

  - start merging small compatible changes 
  - merge compatible test classes

**Comment by alexander233** [comment]
@ioggstream Sorry, but at this point I don't think this is possible. The diff is very large - I have had to change almost every method; for the upgrade to HttpClient4 I kept all HttpClient3 methods as they were (was easier for me to do conversion and debugging); so right now there are both the methods with the old method signature and the new method signature. 
Before merging, we would have to decide how to approach this duplicity of methods (keep or delete?)
Because of the many changes which I have made, rebasing is also a major piece of work. For example, there were many javadoc changes if I recollect correctly. Good improvements, but I don't see how I could incorporate them quickly. 
For me vacations are approaching. Unfortunately I will have very little time for this topic in the next four weeks. Sorry!

**Comment by TheAntimist** [comment]
@alexanderr233 Good stuff there, mate. Glad to see this actually got that far. I'll look it over the upcoming week, sort of busy right now with some things.

Regarding, your work it's preferable if you can follow the take mentioned by @ioggstream . But if that's not possible, I might try slowly merging your work into the project and see where the conflicts occur and fix them as seen fit.

Thanks for all the work regardless.

**Comment by alexander233** [comment]
@ioggstream , @TheAntimist 
Regarding the suggestions of @ioggstream: 
-- Rebase on master: The core problem will be that just merging is not enough. For example, the JavaDoc improvements will probably be merged into the HttpClient3 methods - but they would need to be applied also to the Http4Client methods also. The biggest change in the last few months was the switch to JCR branch. It won't be sufficient to just merge; We may need to copy any changes in Http3Client methods to the associated Http4Client methods, too. 
-- Minimizing the diff: I have not made stylistic changes.. I would just have to remove my TODO comments and other comments that helped me keep track of the progress... 

What I am not clear on is your strategy going forward. Do you want to release Caldav4j in a backward compatible mode and provide simultaneous support for HttpClient3 and HttpClient4?  I am not sure whether that is a good idea. Why burden people who are content with HttpClient3 with all the new classes and jars for HttpClient4? Wouldn't it be better to make another Caldav4j branch supporting only HttpClient4 (or is that your focus anyway)? This would also allow us to make some API changes which bring us better in line with HttpClient4. It would be good to take a look at the API before releasing it...

Maybe I can find some time over the next few days to integrate the switch to JCR branch into my Http4 code. As far as I can see that is the most substantial code change since I pulled master last year. 

- Alexander

**Comment by ioggstream** [comment]
@TheAntimist @alexander233 

  * the -jcr branch worked, so I think we could create an -http4 branch.
  * that means *drop* http3.x and be ready for http4
  * about method factories & co, I was guessing if using some magic (eg. introspection) could simplify things and reduce code. We thus could: shrink http3 codebase, use introspection to switch to http4.

Q: is jcr independent from httpX ?

**Comment by ioggstream** [comment]
@alexander233 @TheAntimist 

  - the ical4j upgrade could be merged if test works
  - the .gitignore upgrade is ok
  - why the servlet downgrade?

**Comment by TheAntimist** [comment]
@alexander233: As @ioggstream stated the plan was to drop http3.x after a final release, and then the next release is to be http4.x only.

@ioggstream The ical4j upgrade should be fine, not sure why the servlet was downgrade is needed.

Also to answer your question jcr is dependent on httpx, to do the brunt of the work.

**Comment by alexander233** [comment]
I have started preparations for rebasing. I will see how much time I can squeeze in for this until next Wednesday.

**Comment by alexander233** [comment]
I have rebased my changes on Caldav4j/master. The core of the changes happens in CalDAVCollection and in HttpXXX method classes. I have not done much on the test classes; but I have added an AdHocTest4.java test class where I ran my unit tests (I tested with Yahoo Calendar). You may want to delete that class later on but it may be helpful to get testing started. 

I ran into one problem with CalDAVCredential: In the constructor method (CaldavCredential(String uri), the port number is set explicitly to 80 or 433 depending on the server scheme if the the port number is unassigned (-1). This did not work for connecting to Yahoo. But if I left the port number undefined (-1) in such cases, the connection worked and tests passed. I haven't tested this extensively, am not sure whether this is the right approach but this change in CalDAVCredential may break httpclient3 calls. 

The whole approach is not perfect, because I just refactored the methods in CalDAVCollection for working with httpclient4; I did not try to adapt the flow or method signatures to the httpclient4 flow. 

The following dependencies need to be added: 
org.apache.httpcomponents:httpclient:4.5.4
org.apache.httpcomponents:httpcore:4.4.8
org.apache.jackrabbit:jackrabbit-webdav:2.13.7
org.threeten:threetenbp1.3.6  //I am not sure what this is used for, might be a transitive dependency

A note on the way I did the rebase: After spending some time trying to go through the git-based rebase process and compare the changes, I gave up, pulled Caldav4j master and then applied my changes on master (and then forced this change onto httpclient-dev). This means that code changes to CalDAVCollection or http methods that happened between November 2017 and now are not included in my translated http4client code. I looked through various committs in that time frame but did not find any such changes. If you know of any such changes, in the last half year, please let know. Going through the merge process was prohibitive because of the many Javdoc related and code-cleanup changes in that time frame and the number of methods that I had to change.

**Comment by TheAntimist** [comment]
Currently merged to local httpclient4 branch from @alexander233 's branch. Work has begun in the merge and reviewing.

**Comment by TheAntimist** [comment]
Pushed initial merge commit at 417a20f5d88a6381ea33d326cfb12918127d1190

**Comment by TheAntimist** [comment]
Migration of main code is done, here 88fc1efa089e36e9286117213d015d6ba3404fe4 . Only thing that remains is the updation of Tests and verifying them.

**Comment by TheAntimist** [comment]
Tests have been migrated in d236d922740412ce66ac0b2ee17d5726d7eb7f1e . Now only the actual testing remains.

**Comment by TheAntimist** [comment]
All tests now work with commit 44ea99d5d3d81e8f0f7f55078a62a873211fd18a .

Although, there's an issue with [`PutGetTest#testPutnonLatin`](https://github.com/caldav4j/caldav4j/blob/http4client-dev/src/test/java/org/osaf/caldav4j/methods/PutGetTest.java#L105) , when running with `mvn test`, but works fine individually, could be a Windows issue, will check back on Linux.

**Comment by TheAntimist** [comment]
Fixed the issue, through a99b278bfb7a95933e77ba49654c0c4730709fb7 . Enforced setting of file encoding for Surefire Plugin, due to an outstanding issue: https://issues.apache.org/jira/browse/SUREFIRE-951

**Comment by TheAntimist** [comment]
@ioggstream In [Line 353, of CalDAVCollection](https://github.com/caldav4j/caldav4j/blob/master/src/main/java/org/osaf/caldav4j/CalDAVCollection.java#L353), We retry three times with different UID's.

In HttpClient4, it already attempts to retry the request 3 times, by default, thus, on each try in this case, on a worst case, we would get 9 tries. Can we shorten this?

**Comment by alexander233** [comment]
@Antimist, @ioggstream : The for loop does not work as intended. Contrary to what the comment says, the loop is not executed 3 times on PRECONDITION_FAILED. The switch statement at the end of the for loop calls MethodUtil.statusToException() if status code is different from SC_CREATED or SC_NO_CONTENT which then throws an uncaught ResourceOutOfDateException if status code is PRECONDITION_FAILED (and other exceptions on all other error codes above 300 which means that it does not retry at all). 
Either add another case for SC_PRECONDITION_FAILED (going back into the loop only into exactly this case) or eliminate the loop. The first alternative seems better because it handles the rare but possible conflict when a generated UID matches an already existing UID.

**Comment by TheAntimist** [comment]
Oh, good catch there, @alexander233 . Yes, I was trying to get that fixed, that code seemed to be some old crux which wasn't properly checked. I'll have to shorten this / fix it soon then.

**Comment by ioggstream** [comment]
@alexander233 Thanks for noting that up!
 Can you please create a testcase and make a PR? It would really be great!

**Comment by ioggstream** [comment]
@TheAntimist the PRECONDITION_FAILED should be related to conflicting UID so I'd preserve the for-loop. We could then limit the for-loop to the case when UID is assigned by caldav4j and not if it's passed by the user. What do you think?

**Comment by alexander233** [comment]
@ioggstream Unfortunately I don't have the Caldav4j test infrastructure so I can't write a test case.

**Comment by TheAntimist** [comment]
@ioggstream Seems fine to me. As far as I see, @alexander233's solution to add a switch condition sounds good.

@alexander233, you could manually setup something like Baikal, otherwise, I'll make the edit if need be.

**Comment by TheAntimist** [comment]
@ioggstream Fixed the retry issue in 7a6365790a50372d29ed81c1d7a3df656e87d259 . Also, added a testcase for verification in CalDAVCollectionTest.

**Comment by TheAntimist** [comment]
@ioggstream Is there any other issues that we are handling for HttpClient 4.x migration? I seem to have fixed everything I can find or get my hands on.

**Comment by TheAntimist** [comment]
Fix merged with #89
